staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go

```diff
@@ -109,15 +129,69 @@ func (o *ResourceConfig) versionEnabled(version schema.GroupVersion) bool {
 	return enabled
 }
 
+// apiAvailable compares the APILifecycle against the emulationVersion.
+// An API is unavailable if it introduced after the emulationVersion,
+// or removed before the emulationVersion.
+func (o *ResourceConfig) apiAvailable(lifecycle schema.APILifecycle) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EmulationVersion) {
+		return true, nil
+	}
+	return isAPIAvailable(lifecycle, o.emulationVersion)
+}
+
+func isAPIAvailable(lifecycle schema.APILifecycle, emulationVersion *version.Version) (bool, error) {
+	// emulationVersion is not set.
+	if emulationVersion.Major() == 0 && emulationVersion.Minor() == 0 {
+		return true, nil
+	}
+	// GroupVersion is introduced after the emulationVersion.
+	if lifecycle.IntroducedVersion != nil && emulationVersion.LessThan(lifecycle.IntroducedVersion) {
+		return false, fmt.Errorf("api introduced at %s, after the emulationVersion %s", lifecycle.IntroducedVersion.String(), emulationVersion.String())
+	}
+	// GroupVersion is removed before the emulationVersion.
+	if lifecycle.RemovedVersion != nil && emulationVersion.GreaterThan(lifecycle.RemovedVersion) {
+		return false, fmt.Errorf("api removed at %s, before the emulationVersion %s", lifecycle.RemovedVersion.String(), emulationVersion.String())
+	}
+	// TODO: handle remove when RemovedVersion equals emulationVersion like resourceExpirationEvaluator.ShouldServeForVersion.
+	return true, nil
+}
+
+// versionAvailable checks if a GroupVersion is available based on its APILifecycle.
+func (o *ResourceConfig) versionAvailable(version schema.GroupVersion) (bool, error) {
+	if o.lifecycleRegistry == nil {
+		return true, nil
+	}
+	return o.apiAvailable(o.lifecycleRegistry.GroupVersionLifecycle(version))
+}
+
+// versionAvailable checks if a GroupVersionResource is available based on its APILifecycle.
+func (o *ResourceConfig) resourceAvailable(resource schema.GroupVersionResource) (bool, error) {
+	if o.lifecycleRegistry == nil {
+		return true, nil
+	}
+	// both the group version and resource have to avaible.
+	if available, err := o.apiAvailable(o.lifecycleRegistry.ResourceLifecycle(resource)); !available {
```

Comment:
Since this makes `--emulated-version=1.N` followed by `--emulated-version=1.N-1`, what is our advice to consumers regarding namespace deletion after going to 1.N?